src/theft_rng.c

```diff
@@ -1,152 +1,55 @@
-   M. Matsumoto and T. Nishimura,
-     ``Mersenne Twister: a 623-dimensionally equidistributed
-       uniform pseudorandom number generator''
-     ACM Transactions on Modeling and 
-     Computer Simulation 8. (Jan. 1998) 3--30.
-
-   Any feedback is very welcome.
-   http://www.math.hiroshima-u.ac.jp/~m-mat/MT/emt.html
-   email: m-mat @ math.sci.hiroshima-u.ac.jp (remove spaces)
-*/
-
-/* The code has been modified to store internal state in heap/stack
- * allocated memory, rather than statically allocated memory, to allow
- * multiple instances running in the same address space.
- * 
- * Also, the functions in the module's public interface have
- * been prefixed with "theft_rng_". */
+/*
+ * Copyright (c) 2017, Koz Ross <koz.ross@retro-freedom.nz>
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
 
 #include <stdlib.h>
-#include <stdio.h>
 #include "theft_rng.h"
 
-#define THEFT_MT_PARAM_N 312
 struct theft_rng {
-    uint64_t mt[THEFT_MT_PARAM_N]; /* the array for the state vector  */
-    int16_t mti;
+  uint64_t state[2];
 };
 
-#define NN THEFT_MT_PARAM_N
-#define MM 156
-#define MATRIX_A 0xB5026F5AA96619E9ULL
-#define UM 0xFFFFFFFF80000000ULL /* Most significant 33 bits */
-#define LM 0x7FFFFFFFULL /* Least significant 31 bits */
-
-static uint64_t genrand64_int64(struct theft_rng *r);
-
-/* Heap-allocate a mersenne twister struct. */
 struct theft_rng *theft_rng_init(uint64_t seed) {
-    struct theft_rng *mt = malloc(sizeof(struct theft_rng));
-    if (mt == NULL) { return NULL; }
-    theft_rng_reset(mt, seed);
-    return mt;
+  struct theft_rng *rng = malloc(sizeof(struct theft_rng));
+  theft_rng_reset(rng, seed);
+  return rng;
 }
 
-/* Free a heap-allocated mersenne twister struct. */
-void theft_rng_free(struct theft_rng *mt) {
-    free(mt);
+void theft_rng_free(struct theft_rng *rng) {
+  free(rng);
 }
 
-/* initializes mt[NN] with a seed */
-void theft_rng_reset(struct theft_rng *mt, uint64_t seed)
-{
-    mt->mt[0] = seed;
-    uint16_t mti = 0;
-    for (mti=1; mti<NN; mti++) {
-        mt->mt[mti] = (6364136223846793005ULL *
-            (mt->mt[mti-1] ^ (mt->mt[mti-1] >> 62)) + mti);
-    }
-    mt->mti = mti;
+void theft_rng_reset(struct theft_rng *rng, uint64_t seed) {
+  uint64_t candidate_state[2] = { seed, ~seed };
```

Comment:
The intent behind `seed, ~seed` is to avoid long runs of zero bits, correct? ("Ensures that bad seeds don't affect the generator")